lisa/util/parallel.py

```diff
@@ -2,29 +2,76 @@
 # Licensed under the MIT license.
 
 from concurrent.futures import FIRST_COMPLETED, Future, ThreadPoolExecutor, wait
-from typing import Any, Callable, Generic, List, Optional, TypeVar
+from typing import Any, Callable, Dict, Generic, List, Optional, TypeVar
+
+from lisa.util.logger import Logger, get_logger
+from lisa.util.perf_timer import create_timer
 
 from . import LisaException
 
 T_RESULT = TypeVar("T_RESULT")
 
 
+class Task(Generic[T_RESULT]):
+    def __init__(
+        self,
+        task_id: int,
+        task: Callable[[], T_RESULT],
+        parent_logger: Optional[Logger],
+    ) -> None:
+        self.id = task_id
+        self._task = task
+        self._lifecycle_timer = create_timer()
+        self._wait_timer = create_timer()
+        self._log = get_logger("Task", str(self.id), parent_logger)
+        self._log.debug(f"Generate task: {self}")
+
+    def close(self) -> None:
+        self._lifecycle_time = self._lifecycle_timer.elapsed_text()
+        self._log.debug(
+            f"Task finished. "
+            f"Lifecycle time: {self._lifecycle_time} "
+            f"Wait time before call: {self._wait_time}"
+            f"Call time: {self._call_time}"
+        )
+
+    def __call__(self) -> T_RESULT:
+        self._call_timer = create_timer()
+        self._wait_time = self._wait_timer.elapsed_text()
```

Comment:
It can be called like `self._wait_timer.elapsed()`. It will record the current time. After that, when it's called like `self._wait_timer.elapsed_text()` again, it's the same time. So it doesn't need to save the string in a variable.